src/cursor/cur_version.c

```diff
@@ -0,0 +1,383 @@
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    table_cursor->set_key(table_cursor);
+}
+
+/*
+ * __curversion_get_key --
+ *     WT_CURSOR->get_key implementation for version cursors.
+ */
+static int
+__curversion_get_key(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    va_list ap;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    va_start(ap, cursor);
+    WT_ERR(__wt_cursor_get_keyv(cursor, cursor->flags, ap));
+    WT_ERR(__wt_cursor_get_keyv(table_cursor, table_cursor->flags, ap));
+
+err:
+    va_end(ap);
+    return (ret);
+}
+
+/*
+ * __curversion_next --
+ *     WT_CURSOR->next method for version cursors. The next function will position the cursor on the
+ *     next update of the key it is positioned at. We traverse through updates on the update chain,
+ *     then the ondisk value, and finally from the history store.
+ */
+static int
+__curversion_next(WT_CURSOR *cursor)
+{
+    WT_CURSOR *hs_cursor, *table_cursor;
+    WT_CURSOR_BTREE *cbt;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    WT_ITEM *hs_key, *hs_value;
+    WT_SESSION_IMPL *session;
+    WT_TIME_WINDOW tw;
+    WT_UPDATE *upd;
+    uint32_t hs_btree_id, hs_upd_type;
+    uint64_t hs_counter, ondisk_stop_txn;
+    wt_timestamp_t ondisk_durable_stop_ts;
+    wt_timestamp_t ondisk_stop_ts;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    hs_cursor = version_cursor->hs_cursor;
+    table_cursor = version_cursor->table_cursor;
+    cbt = (WT_CURSOR_BTREE *)table_cursor;
+    CURSOR_API_CALL(cursor, session, next, NULL);
+
+    ret = WT_NOTFOUND;
+
+    /* The cursor should be positioned. */
+    if (!F_ISSET(cursor, WT_CURSTD_KEY_INT))
+        goto err;
+
+    upd = version_cursor->next_upd;
+
+    if (upd != NULL && !F_ISSET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED)) {
+        /*
+         * If the update is an aborted update, we want to skip to the next update immediately or get
+         * the ondisk value if the update is the last one in the update chain.
+         */
+        while (upd != NULL && upd->txnid == WT_TXN_ABORTED) {
+            upd = upd->next;
+        }
+
+        if (upd == NULL)
+            F_SET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED);
+        else {
+            /*
+             * If the update is a tombstone, we still want to record the stop information but we
+             * also need traverse to the next update to get the full value. If the tombstone was the
+             * last update in the update list, retrieve the ondisk value.
+             */
+            version_cursor->upd_txnid = upd->txnid;
+            version_cursor->upd_durable_stop_ts = upd->durable_ts;
+            version_cursor->upd_stop_ts = upd->start_ts;
+            if (upd->type == WT_UPDATE_TOMBSTONE)
+                upd = upd->next;
```

Comment:
Still the next update after the tombstone may be aborted. We should do
while (upd != NULL && upd->txnid == WT_TXN_ABORTED)
            upd = upd->next;